src/main/java/org/junit/internal/StackTraces.java

```diff
@@ -0,0 +1,203 @@
+            List<String> stackTraceLines, boolean hasCause) {
+        State state = State.PROCESSING_OTHER_CODE;
+        int linesToInclude = stackTraceLines.size();
+        for (String stackTraceLine : asReversedList(stackTraceLines)) {
+            state = state.processLine(stackTraceLine);
+            if (state == State.DONE) {
+                List<String> trimmedLines = stackTraceLines.subList(0, linesToInclude);
+                if (!hasCause) {
+                    return trimmedLines;
+                }
+                List<String> copy = new ArrayList<String>(trimmedLines);
+                copy.add("\t..." + (stackTraceLines.size() - copy.size()) + "more");
+                return copy;
+            }
+            linesToInclude--;
+        }
+        return Collections.emptyList();
+    }
+
+    private static <T> List<T> asReversedList(final List<T> list) {
+        return new AbstractList<T>() {
+
+            @Override
+            public T get(int index) {
+                return list.get(list.size() - index - 1);
+            }
+
+            @Override
+            public int size() {
+                return list.size();
+            }
+        };
+    }
+
+    private enum State {
+        PROCESSING_OTHER_CODE {
+            @Override public State processLine(String line) {
+                if (isTestFrameworkStackTraceLine(line)) {
+                    return PROCESSING_TEST_FRAMEWORK_CODE;
+                }
+                return this;
+            }
+        },
+        PROCESSING_TEST_FRAMEWORK_CODE {
+            @Override public State processLine(String line) {
+                if (isReflectionStackTraceLine(line)) {
+                    return PROCESSING_REFLECTION_CODE;
+                } else if (isTestFrameworkStackTraceLine(line)) {
+                    return this;
+                }
+                return PROCESSING_OTHER_CODE;
+            } 
+        },
+        PROCESSING_REFLECTION_CODE {
+            @Override public State processLine(String line) {
+                if (isReflectionStackTraceLine(line)) {
+                    return this;
+                } else if (isTestFrameworkStackTraceLine(line)) {
+                    // This is here to handle TestCase.runBare() calling TestCase.runTest().
+                    return PROCESSING_TEST_FRAMEWORK_CODE;
+                }
+                return DONE;
+            } 
+        },
+        DONE {
+            @Override public State processLine(String line) {
+                return this;
+            } 
+        };
+
+        /** Processes a stack trace line, possibly moving to a new state. */
+        public abstract State processLine(String line);
+    }
+
+    private static final String[] TEST_FRAMEWORK_METHOD_NAME_PREFIXES = {
+        "org.junit.runner.",
+        "org.junit.runners.",
+        "org.junit.experimental.runners.",
+        "org.junit.internal.",
+        "junit.",
+        "org.apache.maven.surefire."
```

Comment:
Here is the output from a Gradle test failure. I don't use Gradle myself, so I asked a colleague.

````
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="com.example.ImmutableClassTest" tests="2" skipped="0" failures="1" errors="0" timestamp="2015-05-07T16:27:39" hostname="retinahandle.local" time="0.164">
  <properties/>
  <testcase name="how_I_think_it_should_work" classname="com.example.ImmutableClassTest" time="0.163">
    <failure message="org.mutabilitydetector.unittesting.MutabilityAssertionError: &#10;Expected: com.example.ImmutableClass to be IMMUTABLE&#10;     but: com.example.ImmutableClass is actually NOT_IMMUTABLE&#10;    Reasons:&#10;        Field can have a mutable type (java.util.Optional) assigned to it. [Field: foo, Class: com.example.ImmutableClass]&#10;        Field can have a mutable type (java.util.Optional) assigned to it. [Field: bar, Class: com.example.ImmutableClass]&#10;    Allowed reasons:&#10;        Field can have collection with mutable element type (java.util.Optional&lt;com.example.ImmutableClass$Bar&gt;) assigned to it. [Field: bar, Class: com.example.ImmutableClass]&#10;" type="org.mutabilitydetector.unittesting.MutabilityAssertionError">org.mutabilitydetector.unittesting.MutabilityAssertionError: 
Expected: com.example.ImmutableClass to be IMMUTABLE
     but: com.example.ImmutableClass is actually NOT_IMMUTABLE
    Reasons:
        Field can have a mutable type (java.util.Optional) assigned to it. [Field: foo, Class: com.example.ImmutableClass]
        Field can have a mutable type (java.util.Optional) assigned to it. [Field: bar, Class: com.example.ImmutableClass]
    Allowed reasons:
        Field can have collection with mutable element type (java.util.Optional&lt;com.example.ImmutableClass$Bar&gt;) assigned to it. [Field: bar, Class: com.example.ImmutableClass]

	at org.mutabilitydetector.unittesting.internal.AssertionReporter.assertThat(AssertionReporter.java:48)
	at org.mutabilitydetector.unittesting.MutabilityAsserter.assertInstancesOf(MutabilityAsserter.java:125)
	at org.mutabilitydetector.unittesting.MutabilityAssert.assertInstancesOf(MutabilityAssert.java:741)
	at com.example.ImmutableClassTest.how_I_think_it_should_work(ImmutableClassTest.java:24)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:49)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
	at org.gradle.internal.concurrent.DefaultExecutorFactory$StoppableExecutorImpl$1.run(DefaultExecutorFactory.java:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
</failure>
  </testcase>
  <testcase name="this_works_now" classname="com.example.ImmutableClassTest" time="0.001"/>
  <system-out><![CDATA[]]></system-out>
  <system-err><![CDATA[]]></system-err>
</testsuite>

````